contracts/kernel/Kernel.sol

```diff
@@ -5,227 +5,52 @@ import "./KernelStorage.sol";
-    * @param _role Identifier for a group of actions in app
-    * @return address of the manager for the permission
-    */
-    function getPermissionManager(address _app, bytes32 _role) view public returns (address) {
-        return permissionManager[_app][_role];
+    function acl() view public returns (IACL) {
+        return IACL(getApp(ACL_APP));
     }
 
     /**
-    * @dev Function called by apps to check ACL on kernel or to check permission statu
-    * @param _entity Sender of the original call
-    * @param _app Address of the app
-    * @param _role Identifier for a group of actions in app
+    * @dev Function called by apps to check ACL on kernel or to check permission status
+    * @param who Sender of the original call
+    * @param where Address of the app
+    * @param what Identifier for a group of actions in app
+    * @param how Extra data for ACL auth
     * @return boolean indicating whether the ACL allows the role or not
     */
-    function hasPermission(address _entity, address _app, bytes32 _role) view public returns (bool) {
-        return permissions[_entity][_app][_role];
-    }
-
-    /**
-    * @dev Function called by AppProxies to get app code reference
-    * @param _appId Identifier for app
-    * @return address for app code
-    */
-    function getAppCode(bytes32 _appId) view public returns (address) {
-        return appCode[_appId];
-    }
-
-    /**
-    * @dev Internal createPermission for access inside the kernel (on instantiation)
-    */
-    function _createPermission(
-        address _entity,
-        address _app,
-        bytes32 _role,
-        address _manager
-    )
-        internal
-    {
-        // only allow permission creation when it has no manager (hasn't been created before)
-        require(permissionManager[_app][_role] == address(0));
-
-        _setPermission(
-            _entity,
-            _app,
-            _role,
-            true
-        );
-        _setPermissionManager(_manager, _app, _role);
-    }
-
-    /**
-    * @dev Internal function called to actually save the permission
-    */
-    function _setPermission(
-        address _entity,
-        address _app,
-        bytes32 _role,
-        bool _allowed
-    )
-        internal
-    {
-        permissions[_entity][_app][_role] = _allowed;
-
-        SetPermission(
-            _entity,
-            _app,
-            _role,
-            _allowed
-        );
-    }
-
-    /**
-    * @dev Internal function that sets management
-    */
-    function _setPermissionManager(address _newManager, address _app, bytes32 _role) internal {
-        require(_newManager != address(0));
-
-        permissionManager[_app][_role] = _newManager;
-        ChangePermissionManager(_app, _role, _newManager);
+    function hasPermission(address who, address where, bytes32 what, bytes how) view public returns (bool) {
+        IACL _acl = acl();
+        return _acl == address(0) ? true : _acl.hasPermission(who, where, what, how);
```

Comment:
I don't like this `_acl == address(0)` business.

We should make it a rule that you shouldn't use a kernel until it's initialized, otherwise "undefined behaviour" happens. `initialize()` should call an internal `_setApp()` to get past any access controls otherwise set on public functions.